### ISSUE TITLE
Performance Issues with LoRA Training in DreamBooth: Slowdowns After Image Generation on Apple M2 Ultra

### ISSUE THREAD

**Issue by azrahello** [ml-explore/mlx-examples#1056]
I’ve encountered an issue after several attempts. While training a LoRA model works with flux from mlx-examples, it doesn’t work with other tools, like ComfyUI or mflux. Moreover, none of the LoRA models trained using dreambooth.py seem to yield good results.

```
@Alessandros-Mac-Studio flux % python txt2image.py --model dev --n-images 1 \
--image-size 768x512 --steps 20 \
--seed 1010 \
--adapter ../../ComfyUI/models/loras/FLUX/artfusion-lora-r128-v1.safetensors  \ 
--output test_other_lora.png \
'Smolly3 in studio'
Traceback (most recent call last):
  File "/Volumes/NewHome/mlx-examples/flux/txt2image.py", line 72, in <module>
    load_adapter(flux, args.adapter, fuse=args.fuse_adapter)
  File "/Volumes/NewHome/mlx-examples/flux/txt2image.py", line 34, in load_adapter
    rank = int(lora_config["lora_rank"])
               ~~~~~~~~~~~^^^^^^^^^^^^^
KeyError: 'lora_rank'
```
Additionally, during the iterations, generating images causes a behavior in RAM such that, after the second image generation, the process becomes drastically slower compared to the iterations before the image generation.

I solved the problem by using progress-every steps, setting the number of steps equal to the total number of iterations, which prevents intermediate image generations. However, this approach doesn’t allow me to visually evaluate what I’m creating during training. I’m working on an Apple M2 Ultra with 64 GB of RAM.

While monitoring resources through Apple’s Activity Monitor, I noticed that during the initial iterations, memory “dances rhythmically” between app memory and wired memory. After generating an image, compressed memory starts being used, and the process becomes extremely slow.

Maybe I’m doing something wrong, but even following DreamBooth guides closely, I haven’t seen any significant improvements.
As you can see, the time after the first image generation increases drastically. The gap from 20:06 to 20:21 is due to some applications being open during the operation, but usually, it completes 50 iterations in about 4 minutes when I’m not making intermediate generations.

![Screenshot 2024-10-20 alle 16 34 05](https://github.com/user-attachments/assets/9416ae7e-acae-4517-909d-f36b2e3fc869)
Here are some of the results, which are obviously not good from the various training sessions (not all of them). Only one came close to something usable. Now, I’m not sure if it’s due to my own mistakes, the dataset (which I changed to see if that was the issue). Any help, suggestions, tips, explanations, or even magic is greatly appreciated. I want to thank all of you anyway, as you are doing incredible things and making giant strides in such a short time.
![Screenshot 2024-10-20 alle 16 43 04](https://github.com/user-attachments/assets/b4ec5643-d4c7-4460-8cd5-898dcfcb937a)

Ps 
Would it be possible to resume training from a saved adapter?

**Comment by kirel** [comment]
What completed this?

**Comment by azrahello** [comment]
> What completed this?

Nothing, I noticed that no one replied, so I thought there might be something wrong with my request. I don’t have the skills to solve these kinds of problems or to write code (I’m a photographer and I study fine arts). I thought that maybe due to my lack of knowledge, I made an incorrect request, or perhaps, since these are just “examples,” it’s not in their interest to go into specific details but rather to show other developers how the mechanisms work, leaving it up to them to develop specific features. I might even be slowing down the developers. I stopped testing this branch of my research and moved on to something else. Before opening this issue, I had started a discussion after seeing the discussion and a post [link to Hugging Face post] (https://huggingface.co/TheLastBen/The_Hound) and [Twitter link] (https://x.com/__TheBen/status/1829554120270987740) about creating LoRAs at specific levels, but I couldn’t figure out how to do it.

PS: By using the values in the readme and a dataset with my own photos, I managed to get some results, but nothing that fully satisfies me.
